File: packages/shared-types/src/memory.ts
export interface Memory {
  id: string;
  workspaceId: string;
  scope: 'global' | 'project' | 'agent' | 'device' | 'session';
  type: 'preference' | 'semantic' | 'procedural' | 'decision' | 'episodic' | 'artifact' | 'warning' | 'case';
  content?: string;
  source: string;
  sourceId?: string;
  agentId?: string;
  deviceId?: string;
  confidence: number;
  visibility: 'private' | 'agent' | 'project' | 'shared';
  status: 'active' | 'deprecated' | 'archived' | 'conflict';
  expiresAt?: string;
  lastUsedAt?: string;
  createdAt: string;
  updatedAt: string;
}

export interface CreateMemoryRequest {
  scope: string;
  type: string;
  content: string;
  source: string;
  agentId?: string;
  confidence?: number;
  visibility?: string;
}

export interface MemorySearchParams {
  query?: string;
  scope?: string;
  type?: string;
  agentId?: string;
  status?: string;
  limit?: number;
}

export interface MemoryCandidate {
  id: string;
  workspaceId: string;
  proposedScope?: string;
  proposedType?: string;
  content?: string;
  reason?: string;
  sourceAgentId?: string;
  sourceDeviceId?: string;
  sourceArtifactId?: string;
  confidence: number;
  reviewStatus: 'candidate' | 'needs_review' | 'approved' | 'rejected' | 'merged' | 'auto_rejected';
  reviewNotes?: string;
  createdAt: string;
  reviewedAt?: string;
}

export interface SubmitCandidateRequest {
  agentId: string;
  deviceId?: string;
  taskId?: string;
  candidates: CandidateItem[];
}

export interface CandidateItem {
  scope: string;
  type: string;
  content: string;
  reason?: string;
  confidence?: number;
}

File: apps/collector/src/proxy/memory.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MemorySubmitRequest {
    pub agent_id: String,
    pub device_id: Option<String>,
    pub task_id: Option<String>,
    pub candidates: Vec<MemoryCandidate>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MemoryCandidate {
    pub scope: String,
    pub r#type: String,
    pub content: String,
    pub reason: Option<String>,
    pub confidence: Option<f64>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MemorySubmitResponse {
    pub accepted: Vec<AcceptedCandidate>,
    pub rejected: Vec<RejectedCandidate>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AcceptedCandidate {
    pub candidate_id: String,
    pub review_status: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RejectedCandidate {
    pub content: String,
    pub reason: String,
}
